src/hooks/progress/use-tutorial-progress-mutations/index.ts

```diff
@@ -0,0 +1,123 @@
+ * mutation functions.
+ *
+ * Each of these functions accepts a single object argument:
+ * `{ tutorialId, collectionId, completePercent }` (and, optionally, `options`)
+ * and creates or updates the tutorial with the provided completePercent.
+ */
+const useTutorialProgressMutations = (): UseTutorialProgressMutationsResult => {
+	const queryClient = useQueryClient()
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+
+	/**
+	 * When a mutation is successful, we set the associated query data.
+	 * TODO: once there are other queries, we'll likely want to invalidate those.
+	 */
+	const makeOnMutationSuccess = () => {
+		return (
+			_result: ApiCollectionTutorialProgress,
+			mutationVariables: TutorialProgressMutationVariables
+		) => {
+			// Destructure the variables we need to update related queries
+			const { tutorialId, collectionId, completePercent } = mutationVariables
+			// Update query data for this specific tutorial-in-collection progress
+			const progressLabel = progressPercentToLabel(completePercent)
+			queryClient.setQueryData(
+				[TUTORIAL_PROGRESS_QUERY_ID, tutorialId, collectionId],
+				progressLabel
+			)
+		}
+	}
+
+	/**
+	 * Mutation to create tutorial progress.
+	 */
+	const createTutorialProgressMutation = useMutation(
+		createTutorialProgressApi,
+		{ onSuccess: makeOnMutationSuccess() }
+	)
+	/**
+	 * Function to create tutorial progress.
+	 *
+	 * Note: The createTutorialProgress function needs to update when the
+	 * accessToken or onSuccess option of createTutorialProgressMutation changes.
+	 * We wrap this in useCallback to prevent more updates than necessary.
+	 */
+	const createTutorialProgress = useCallback(
+		(args: TutorialProgressMutationArgs) => {
+			const { tutorialId, collectionId, completePercent, options } = args
+			createTutorialProgressMutation.mutate(
+				{ accessToken, tutorialId, collectionId, completePercent },
+				options
+			)
+		},
+		[createTutorialProgressMutation, accessToken]
+	)
+
+	/**
+	 * Mutation to update tutorial progress.
+	 */
+	const updateTutorialProgressMutation = useMutation(
+		updateTutorialProgressApi,
+		{
+			onSuccess: makeOnMutationSuccess(),
+		}
+	)
+	/**
+	 * Function to update tutorial progress.
+	 *
+	 * Note: as with createTutorialProgress,
+	 * we wrap this in useCallback to prevent more updates than necessary.
+	 */
+	const updateTutorialProgress = useCallback(
+		(args: TutorialProgressMutationArgs) => {
+			const { tutorialId, collectionId, completePercent, options } = args
+
+			updateTutorialProgressMutation.mutate(
+				{ accessToken, tutorialId, collectionId, completePercent },
+				options
+			)
+		},
+		[updateTutorialProgressMutation, accessToken]
+	)
+
+	/**
+	 * Consumers likely only want to perform calculations in preparation to
+	 * run these mutations if status is "idle" or "success". Otherwise,
+	 * attempt to run the mutation may result in a continuous loop.
+	 */
```

Comment:
Where is this happening that `isReady` is needed?

Reply:
It's being used in `use-update-tutorial-progress`- we don't want to continually re-calculate progress and re-do mutations if we have in progress.